Honest verifier zk update

hansieodendaal opened the pull request:
Updates as suggested by Sarang Noether

hansieodendaal commented:
Hi @anselld, please only verify the changes editorially. Thanks

anselld commented on src/cryptography/bulletproofs-protocols/MainReport.md:
```diff
@@ -764,6 +764,11 @@ communication [[29]]:
 Module bulletproofs::range_proof_mpc">29</a>]</b></div>
 
+With the standard MPC protocol implementation as depicted in Figure&nbsp;7, there's no guarantee of Honest Verifier ZK, 
+that we can expect that the dealer behaves honestly according to the specified protocol ([[30]], [[31]]). There is no way to 
```
Meaning of first sentence is unclear.

hansieodendaal replied:
Please elaborate? I do not understand what is unclear.

anselld replied:
The following would be clear to me:
With the standard MPS protocol implementation as depicted in Figure 7, there's no guarantee that we can expect the dealer to behave honestly according to the specified protocol...

I don't understand how 'Honest Verifier ZK' fits in.

hansieodendaal replied:
`Honest Verifier ZK` means it is only ZK if the verifier (dealer in this case) is honest.

hansieodendaal replied:
Maybe: "_With the standard MPC protocol implementation as depicted in Figure&nbsp;7, there's no guarantee of Honest Verifier Zero-knowledge (ZK),  which would mean that the dealer behaves honestly according to the specified protocol ([[30]], [[31]])._"

anselld replied:
Thanks, Hansie - this is fine.

anselld commented on src/cryptography/bulletproofs-protocols/MainReport.md:
```diff
@@ -764,6 +764,11 @@ communication [[29]]:
+With the standard MPC protocol implementation as depicted in Figure&nbsp;7, there's no guarantee of Honest Verifier ZK, 
+that we can expect that the dealer behaves honestly according to the specified protocol ([[30]], [[31]]). There is no way to 
+detect a dishonest challenge until the dealer provides the final proof to the parties, except if every party acts as a 
+dealer in the final step of the protocol. Each party need to ensure that their partial proof elements (ProofShare) reach 
```
Should be "each party needs"

hansieodendaal replied:
```suggestion
dealer in the final step of the protocol. Each party needs to ensure that their partial proof elements (ProofShare) reach 
```

anselld commented on src/cryptography/bulletproofs-protocols/MainReport.md:
```diff
@@ -1251,6 +1259,20 @@ July 2018"
 "Dalek Cryptography - 
 Module bulletproofs::range_proof_mpc"
 
+[[30]] "What is the difference between honest verifier zero knowledge and zero knowledge?" [online]. Available: 
```
Please replace [online]. with [Online.]

hansieodendaal replied:
All references in this report make use of `[online]`, not `[Online]`. Also as per the [Sty;e Guide](https://tlu.tarilabs.com/preface/style-guide.html#markdown-links)

anselld replied:
It depends on the sentence. [Online.] is standalone in this context.

hansieodendaal replied:
I suggest we create a separate PR to change all `[online]` to `[Online]` throughout the TLU content.

anselld replied:
It is not always [Online.] It depends on the context. In the right context, '[online].' is correct.

anselld replied:
For example:
[8] The Beam Team, "GitHub: beam/node/laser_beam_demo at master - BeamMW/beam" [online]. Available: https://github.com/BeamMW/beam/tree/master/node/laser_beam_demo. Date accessed: 2019‑07‑05.

anselld replied:
Hi @hansieodendaal, I cannot see if [online]. has been updated to [Online.}

anselld commented on src/protocols/mimblewimble-mp-bp-utxo/MainReport.md:
```diff
@@ -456,8 +456,8 @@ $ C\_m $.
-| Information sharing security  | Use of Merlin transcripts, combined with the MPC protocol, makes this method more secure against replay attacks. | No specific sharing protocol suggested, but potentially easy to implement, as described [here]((#simple-sharing-protocol)).  |
-| Guarantee of honest verifier  | There's no way to detect a dishonest challenge until the dealer provides the final proof to the players, except if every player acts as a dealer. | Similar  |
+| Information sharing security  | Use of Merlin transcripts, combined with the MPC protocol, makes this method more secure against replay attacks. | No specific sharing protocol suggested, but potentially easy to implement, as described [here](#simple-sharing-protocol).  |
+| Guarantee of honest verifier  | With the standard [MPC protocol implementation](../../cryptography/bulletproofs-protocols/MainReport.md#mpc-protocol-for-bulletproofs), there's no guarantee of Honest Verifier ZK, but could be improved as suggested. | Similar  |
```
Meaning of this sentence is unclear.

hansieodendaal replied:
Please elaborate? I do not understand what is unclear.

anselld replied:
What could be improved? This is unclear. This could be improved? Unclear.

hansieodendaal replied:
Maybe "_With the standard [MPC protocol implementation](../../cryptography/bulletproofs-protocols/MainReport.md#mpc-protocol-for-bulletproofs), there's no guarantee of it being Honest Verifier ZK, but this aspect could be improved as suggested in the report._" ?

anselld replied:
Perfect, thanks.

anselld commented on src/cryptography/bulletproofs-protocols/MainReport.md:
```diff
@@ -1251,6 +1260,20 @@ July 2018"
 "Dalek Cryptography - 
 Module bulletproofs::range_proof_mpc"
 
+[[30]] "What is the difference between honest verifier zero knowledge and zero knowledge?" [Online]. Available: 
```
[Online.] is correct here; [online]. is correct in [31].

hansieodendaal commented:
@anselld, please do a final editing check on the changes? Thanks

hansieodendaal commented:
10:00 <Hansie> Hi sarang, hope you are well. W.r.t. comments you made about the Bulletproofs MPC protocol only being honest verifier zk, I made some adjustments in what I posted for TLU: https://tlu.tarilabs.com/protocols/mimblewimble-mp-bp-utxo/MainReport.html#comparison-of-the-two-bulletproof-methods
10:00  * sarang is away (Auto-away)
10:09 <Hansie> Specifically with evaluation of `Information sharing security` and `Guarantee of honest verifier`. Would you care to review those statements for me please?
15:11 <sarang> Link in the table is a 404 error
18:46 <sarang> One note: you say "... except if every player acts as a dealer"
18:47 <sarang> It may be worth expanding on this: all the players need to do is to ensure their partial proof elements reach every other player
18:47 <sarang> and that each player produce the aggregate challenges identically
18:48 <sarang> Notably, the initial step of assigning numbers/indices to each player can be done by a single distinguished dealer without any trouble, since the ordering of players has nothing to do with security
18:49 <sarang> Also, I seem to recall that the inner product argument _can_ be done by a single dealer without interaction, since the generated challenges are done only with aggregated proof elements
18:49 <sarang> So "every player acts as a dealer" is only partially true
07:15 <Hansie> Thank you sarang, will give it another bash
11:27 <Hansie> Hi Sarang, I implemented the changes. Care to comment on this PR? https://github.com/tari-labs/tari-university/pull/223
16:08 <sarang> The change says that there is no guarantee of it being honest verifier zero knowledge, which seems slightly wrong
16:08 <sarang> The protocol is definitely SHVZK
16:08 <sarang> But the dealer may or may not generate challenges honestly
16:09 <sarang> and the next sentences have it backwards...
16:09 <sarang> players need to be dealers for intermediate challenges that rely on aggregation of partial proof elements, not for the last portion where the inner product compression happens (the input values there are already complete and public)
16:13 <sarang> Perhaps something like this instead: "Since the Bulletproofs protocol is special honest verifier zero knowledge only, a secure multiparty computation requires all players to receive partial proof elements and independently compute aggregated challenges, in order to avoid the alternate case where a single dealer maliciously generates them. A single dealer can, however, assign index positions to all players at the start of 
16:13 <sarang> the protocol, and may complete the inner product compression since it does not rely on partial proof data."
21:35 <sarang> Oh, and lest I forget... it is true that SHVZK implies witness indistinguishability, but it isn't clear to me what the implications of this would be on the MPC in practice
21:36 <sarang> It could be that there are no practical attacks possible from a malicious dealer
21:36 <sarang> and that witness-indistinguishability is sufficient
14:59 <Hansie> Thank you Sarang. (1) I updated the reports, moved this discussion under a separate heading "MPC Protocol Security Discussion", and also added you as a contributor if you could please verify it. (2) Would you say that this method (https://tlu.tarilabs.com/protocols/mimblewimble-mp-bp-utxo/MainReport.html#utilizing-grins-shared-bulletproof-computation) also suffers from a similar security concern having a malicious dealer? I summarized it by saying "There is no guarantee that the dealer behaves honestly."
16:13 <sarang> If each player does not independently compute challenges themselves, then the dealer could be malicious (with unknown result)
16:15 <sarang> There was also discussion with real_or_random a while back about whether precommitment to partial proof values would also be required
16:16 <sarang> I don't think this affects the zk properties, however
12:10 <Hansie> Thank you Sarang, you helped a lot.

anselld commented on src/cryptography/bulletproofs-protocols/MainReport.md:
```diff
@@ -773,6 +773,20 @@ challenge and sends it to each party, who in turn send their witness to the deal
+With the standard MPC protocol implementation as depicted in Figure&nbsp;7, there's no guarantee that the dealer behaves 
+honestly according to the specified protocol and generates challenges honestly. Since the Bulletproofs protocol is 
+special HVZK ([[30]], [[31]]) only, a secure MPC protocol requires all parties to receive 
+partial proof elements and independently compute aggregated challenges, in order to avoid the alternate case where a 
```
Would rather replace 'alternate' with 'alternative'.

anselld commented on src/cryptography/bulletproofs-protocols/MainReport.md:
```diff
@@ -773,6 +773,20 @@ challenge and sends it to each party, who in turn send their witness to the deal
+special HVZK ([[30]], [[31]]) only, a secure MPC protocol requires all parties to receive 
+partial proof elements and independently compute aggregated challenges, in order to avoid the alternate case where a 
+single dealer maliciously generates them. A single dealer can, however, assign index positions to all parties at the 
+start of the protocol, and may complete the inner product compression since it does not rely on partial proof data.
```
Suggest inserting comma after 'compression'.

anselld commented on src/protocols/mimblewimble-mp-bp-utxo/MainReport.md:
```diff
@@ -456,8 +456,8 @@ $ C\_m $.
-| Information sharing security  | Use of Merlin transcripts, combined with the MPC protocol, makes this method more secure against replay attacks. | No specific sharing protocol suggested, but potentially easy to implement, as described [here]((#simple-sharing-protocol)).  |
-| Guarantee of honest verifier  | There's no way to detect a dishonest challenge until the dealer provides the final proof to the players, except if every player acts as a dealer. | Similar  |
+| Information sharing security  | Use of Merlin transcripts, combined with the MPC protocol, makes this method more secure against replay attacks. | No specific sharing protocol suggested, but potentially easy to implement, as described [here](#simple-sharing-protocol).  |
+| Guarantee of honest dealer | With the standard MPC protocol implementation, there is no guarantee that the dealer behaves honestly according to the protocol and generates challenges honestly, but this aspect could be improved [as suggested1](../../cryptography/bulletproofs-protocols/MainReport.md#mpc-protocol-security-discussion) in that report. | Although each party independently compute chall ...
```
```suggestion
| Guarantee of honest dealer | With the standard MPC protocol implementation, there is no guarantee that the dealer behaves honestly according to the protocol and generates challenges honestly, but this aspect could be improved [as suggested1](../../cryptography/bulletproofs-protocols/MainReport.md#mpc-protocol-security-discussion) in that report. | Although each party independently computes challenges themselves, it isn't clear what the implications of a dishonest dealer would be in practice.  |
```

anselld commented:
Hi @hansieodendaal, these look fine; please see suggested changes. Also please search and correct 'et al' with 'et al.'
